kong/clustering/control_plane.lua

```diff
@@ -105,27 +108,179 @@ function _M.new(parent)
   }
 
-  return setmetatable(self, {
+  setmetatable(self, {
     __index = function(tab, key)
       return _M[key] or parent[key]
     end,
   })
+
+  self.plugins_map = {}
+
+  return self
 end
 
 
+local function invalidate_keys_from_config(config_plugins, keys)
+  if not config_plugins then
+    return
+  end
+
+  for _, t in ipairs(config_plugins) do
+    if t and t["config"] then
+      local config = t["config"]
+      local name = gsub(t["name"], "-", "_")
+
+      -- Handle Redis configurations (regardless of plugin)
+      if config.redis then
+        local config_plugin_redis = t["config"].redis
+        for _, key in ipairs(keys["redis"]) do
+          if config_plugin_redis[key] ~= nil then
+            config_plugin_redis[key] = nil
+          end
+        end
+      end
+
+      -- Handle fields in specific plugins
+      if keys[name] ~= nil then
+        for _, key in ipairs(keys[name]) do
+          if config[key] ~= nil then
+            config[key] = nil
+          end
+        end
+      end
+    end
+  end
+end
+
+local function dp_version_num(dp_version)
+  local base = 1000000000
+  local version_num = 0
+  for _, v in ipairs(utils.split(dp_version, ".", 4)) do
+    v = v:match("^(%d+)")
+    version_num = version_num + base * tonumber(v, 10) or 0
+    base = base / 1000
+  end
+
+  return version_num
+end
+-- for test
+_M._dp_version_num = dp_version_num
+
+local removal_fields = {
+  [2003003003] = {
+    file_log = {
+      "custom_fields_by_lua",
+    },
+    http_log = {
+      "custom_fields_by_lua",
+    },
+    loggly = {
+      "custom_fields_by_lua",
+    },
+    prometheus = {
+      "per_consumer",
+    },
+    syslog = {
+      "custom_fields_by_lua",
+    },
+    tcp_log = {
+      "custom_fields_by_lua",
+    },
+    udp_log = {
+      "custom_fields_by_lua",
+    },
+    zipkin = {
+      "tags_header",
+    },
+  },
+
+  [2004001002] = {
+    redis = {
+      "connect_timeout",
+      "keepalive_backlog",
+      "keepalive_pool_size",
+      "read_timeout",
+      "send_timeout",
+    },
+    syslog = {
+      "facility",
+    },
+  }
+}
+
+local function get_removed_fields(dp_version_number)
+  local unknown_fields = {}
+  local has_fields
+
+  -- Merge dataplane unknown fields; if needed based on DP version
+  for v, list in pairs(removal_fields) do
+    if dp_version_number < v then
+      has_fields = true
+      for plugin, fields in pairs(list) do
+        if not unknown_fields[plugin] then
+          unknown_fields[plugin] = {}
+        end
+        for _, k in ipairs(fields) do
+          table.insert(unknown_fields[plugin], k)
+        end
+      end
+    end
+  end
+
+  return has_fields and unknown_fields or nil
+end
+-- for test
+_M._get_removed_fields = get_removed_fields
+
+local function update_compatible_payload(payload, dp_version, log_suffix)
+  local fields = get_removed_fields(dp_version_num(dp_version))
+  if fields then
+    local inflated_payload, err = inflate_gzip(payload)
```

Comment:
We should probably restructure the code so that it does not perform this double inflate/deflate. It could get very expensive quickly (especially with large payloads).

Reply:
this function has been changed to only return deflated payload if it's been changed.
otherwise the called function must used the `deflated_reconfigure_payload`.